Userland/Libraries/LibWeb/Streams/QueueOperations.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2023, Matthew Olsson <mattco@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibJS/Runtime/Error.h>
+
+namespace Web::Streams {
+
+// https://streams.spec.whatwg.org/#value-with-size
+struct ValueWithSize {
+    JS::Value value;
+    double size;
+};
+
+// https://streams.spec.whatwg.org/#dequeue-value
+template<typename T>
+JS::Value dequeue_value(T& container)
+{
+    // 1. Assert: container has [[queue]] and [[queueTotalSize]] internal slots.
+
+    // 2. Assert: container.[[queue]] is not empty.
+    VERIFY(!container.queue().is_empty());
+
+    // 3. Let valueWithSize be container.[[queue]][0].
+    // 4. Remove valueWithSize from container.[[queue]].
+    auto value_with_size = container.queue().take_first();
+
+    // 5. Set container.[[queueTotalSize]] to container.[[queueTotalSize]] − valueWithSize’s size.
+    container.set_queue_total_size(container.queue_total_size() - value_with_size.size);
+
+    // 6. If container.[[queueTotalSize]] < 0, set container.[[queueTotalSize]] to 0. (This can occur due to rounding errors.)
+    if (container.queue_total_size() < 0.0)
+        container.set_queue_total_size(0.0);
+
+    // 7. Return valueWithSize’s value.
+    return value_with_size.value;
+}
+
+// https://streams.spec.whatwg.org/#enqueue-value-with-size
+template<typename T>
+JS::Completion enqueue_value_with_size(T& container, JS::Value value, double size)
+{
+    // 1. Assert: container has [[queue]] and [[queueTotalSize]] internal slots.
+
+    // 2. If ! IsNonNegativeNumber(size) is false, throw a RangeError exception.
+    if (size < 0.0)
+        return JS::throw_completion(MUST_OR_THROW_OOM(JS::TypeError::create(container.realm(), "Chunk has negative size"sv)));
```

Comment:
https://github.com/SerenityOS/serenity/blob/master/Documentation/Browser/Patterns.md#webwebidlsimpleexception

Reply:
According to that it can't be turned into a value until it passes through the bindings layer. I need the value for step 4.5.1 of https://streams.spec.whatwg.org/#readable-stream-default-controller-enqueue, unless I'm missing something

Reply:
You can use `throw_dom_exception_if_needed()` (which is what the generated code does, too), like so:

https://github.com/SerenityOS/serenity/blob/df577b457ace4d0162c10141c2530ebaa735b721/Userland/Libraries/LibWeb/Fetch/Body.cpp#L189-L198